### PR TITLE
Add pathSatisfies

### PR DESCRIPTION
**Pull request by gilligan**
pathSatisfies works analogous to propSatisfies but applies
the predicator on a path instead of a property name.

### CODE REVIEW

File: test/pathSatisfies.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  var isPositive = function(n) { return n > 0; };
+
+  it('returns true if the specified object path satisfies the given predicate', function() {
+    eq(R.pathSatisfies(isPositive, ['x', 'y'], {x: {y: 1}}), true);
+  });
+
+  it('returns false otherwise', function() {
+    eq(R.pathSatisfies(isPositive, ['x', 'y'], {x: {y: 0}}), false);
+  });
+
```
Comment (CrossEye) [inline test/pathSatisfies.js]: I think this could use another test demonstrating the behavior on

    R.pathSatisfies(isPositive, ['x', 'y'], {x: {z: 42}}

**Comment by CrossEye** [comment]
This looks reasonable to me.  I think `path` and `prop` are clearly parallel in most everything.

The only question is how far we want to carry this, and how important people find such a function.  Is it useful and important enough for the core, or is it more suited for the Cookbook?

**Comment by gilligan** [comment]
@CrossEye I added it to our code base as a utility function early on and I thought it would make sense to have it as built-in function in Ramda.

Of course you could then begin to wonder if there should also be a `lensPath` analogous to `lensProp` ...

File: test/pathSatisfies.js (hunk 1 above)
Comment (gilligan) [inline test/pathSatisfies.js]: Indeed. Will add it.

File: src/pathSatisfies.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ * @memberOf R
+ * @since v0.18.1
+ * @category Logic
+ * @sig (a -> Boolean) -> [String] -> {String: a} -> Boolean
+ * @param {Function} pred
+ * @param {Array} propPath
+ * @param {*} obj
+ * @return {Boolean}
+ * @see R.propSatisfies
+ * @see R.path
```
Comment (davidchambers) [inline src/pathSatisfies.js]: Multiple `@see` annotations may appear on one line, separated by commas (e.g. `@see R.foo, R.bar`).

File: src/pathSatisfies.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+var path = require('./path');
+
+
+/**
+ * Returns `true` if the specified object property at given path satisfies the given
+ * predicate; `false` otherwise.
+ *
+ * @func
+ * @memberOf R
+ * @since v0.18.1
```
Comment (davidchambers) [inline src/pathSatisfies.js]: Let's remove this line.

File: test/pathSatisfies.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+describe('pathSatisfies', function() {
+
+  var isPositive = function(n) { return n > 0; };
+
+  it('returns true if the specified object path satisfies the given predicate', function() {
+    eq(R.pathSatisfies(isPositive, ['x', 'y'], {x: {y: 1}}), true);
+  });
+
+  it('returns false if the specified path does not exist', function() {
+    R.pathSatisfies(isPositive, ['x', 'y'], {x: {z: 42}});
```
Comment (davidchambers) [inline test/pathSatisfies.js]: There's no assertion here. ;)

Comment (CrossEye) [inline test/pathSatisfies.js]: sez you!

**Comment by gilligan** [comment]
@CrossEye @davidchambers any further objections on this? ;)

File: src/pathSatisfies.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
+var _curry3 = require('./internal/_curry3');
+var path = require('./path');
+
+
+/**
+ * Returns `true` if the specified object property at given path satisfies the given
+ * predicate; `false` otherwise.
```
Comment (davidchambers) [inline src/pathSatisfies.js]: Let's wrap this at 79 characters (see #1461).

File: src/pathSatisfies.js [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+ * @category Logic
+ * @sig (a -> Boolean) -> [String] -> {String: a} -> Boolean
+ * @param {Function} pred
+ * @param {Array} propPath
+ * @param {*} obj
+ * @return {Boolean}
+ * @see R.propSatisfies, R.path
+ * @example
+ *
+ *      R.propSatisfies(y => y > 0, ['x', 'y'], {x: {y: 2}}); //=> true
```
Comment (davidchambers) [inline src/pathSatisfies.js]: `s/prop/path/`

File: src/pathSatisfies.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+
+/**
+ * Returns `true` if the specified object property at given path satisfies the given
+ * predicate; `false` otherwise.
+ *
+ * @func
+ * @memberOf R
+ * @category Logic
+ * @sig (a -> Boolean) -> [String] -> {String: a} -> Boolean
```
Comment (davidchambers) [inline src/pathSatisfies.js]: `{String: a}` doesn't capture nesting. Perhaps `Object` would be better.

File: test/pathSatisfies.js (hunk 1 above)
Comment (davidchambers) [inline test/pathSatisfies.js]: We also need a test for the empty path.

File: src/pathSatisfies.js [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+ * @example
+ *
+ *      R.pathSatisfies(y => y > 0, ['x', 'y'], {x: {y: 2}}); //=> true
+ */
+module.exports = _curry3(function pathSatisfies(pred, propPath, obj) {
+  switch (propPath.length) {
+    case 0:
+      return false;
+    case 1:
+      return pred(obj[propPath[0]]);
```
Comment (davidchambers) [inline src/pathSatisfies.js]: This case is unnecessary. The case below can cover `propPath.length === 1` as well.

**Comment by gilligan** [comment]
@davidchambers thanks for the thorough review, fixed everything accordingly

**Comment by gilligan** [comment]
@davidchambers alright, premature optimization I suppose. Use a simple if/else instead of the switch statement then I guess ?

**Comment by gilligan** [comment]
Or even 
```js
module.exports = _curry3(function pathSatisfies(pred, propPath, obj) {                                                                                         
  return (propPath.length === 0) ? false : pred(path(propPath, obj));                                   
});
```

**Comment by davidchambers** [comment]
Even simpler:

```javascript
propPath.length > 0 && pred(path(propPath, obj))
```

**Comment by gilligan** [comment]
ah, so now tests passed - any other objections to this? :)

File: test/pathSatisfies.js (hunk 1 above)
Comment (davidchambers) [inline test/pathSatisfies.js]: Could you add a test for the empty path, @gilligan?

Comment (CrossEye) [inline test/pathSatisfies.js]: Outside that, this LGTM.

Comment (gilligan) [inline test/pathSatisfies.js]: Sorry, was away - I added the empty path test.

**Comment by CrossEye** [comment]
LGTM.  Anyone else with thoughts?

**Comment by davidchambers** [comment]
:deciduous_tree:
